src/BiDirectionalNode.h

```diff
@@ -0,0 +1,44 @@
+//
+// Created by shvmo on 02/05/2020.
+//
+
+#ifndef DATASTRUCTURESWET1_BIDIRECTIONALNODE_H
+#define DATASTRUCTURESWET1_BIDIRECTIONALNODE_H
+
+#include <memory>
+
+using std::shared_ptr;
+using std::weak_ptr;
+
+template<class Type>
+class BiDirectionalNode {
+public:
+    BiDirectionalNode();
+
+    ~BiDirectionalNode();
+
+    Type &getValue();
+
+    shared_ptr<BiDirectionalNode<Type>> getNext() const;
+
+    // Done for optimization when we want to iterate a bit faster
+    BiDirectionalNode<Type> *getNextPointer() const;
+
+    weak_ptr<BiDirectionalNode<Type>> getPrevious() const;
+
+    void setValue(const Type &newValue);
+
+    void setNext(shared_ptr<BiDirectionalNode<Type>> newNext);
+
+    void setPrevious(shared_ptr<BiDirectionalNode<Type>> newPrevious);
+
+private:
+    Type value;
+    shared_ptr<BiDirectionalNode<Type>> next;
+    // This is of type weak_ptr because otherwise it can create cyclic dependency and create memory leaks if shared_ptr is used
+    weak_ptr<BiDirectionalNode<Type>> previous;
```

Comment:
amazing idea but it seems like we just hide potential misuse of this class 🤔 
i mean, i want to know if we forgot to reassign a new previous.
we don't want to let previous be `nullptr` because of potential errors.

Reply:
Nope it's okay. It knows how to deal with nullptr, and that's the use case you use weak ptr for
https://stackoverflow.com/questions/38298008/possible-memory-leaks-with-smart-pointers/51412891

Reply:
if we can code bi-directional linked list in C i'm sure we can code one in cpp with shared pointers. 
it just means we need to be more careful when we implement the list class and how we use the node class.
so i'm no convinced it's CRUCIAL but i'll take it to heart.